### DISCUSSION TITLE
Problems when change to a new PDK

### DISCUSSION THREAD

**Discussion by 706697916** [The-OpenROAD-Project/OpenROAD-flow-scripts#1060] [General] [answered]
I want to develop a new PDK in OpenROAD. I follow the guid "https://openroad-flow-scripts.readthedocs.io/en/latest/contrib/PlatformBringUp.html". Now I have changed the PDK file, makefile, config file. But it seems that OpenROAD cannot recognize the standard cells in new PDK lib.
The three pictures shows my problem. How to solve it or how to debug the codes?
<img width="1072" alt="1683818319595" src="https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/70256434/cdad7124-8173-4f29-bfc9-c85776645364">
<img width="1110" alt="1683818461249" src="https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/70256434/f4b701e2-52c2-4778-b442-ce0553ced106">
<img width="950" alt="1683818516179" src="https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/70256434/8745135a-3933-45b3-9974-037a04958901">

**Comment by maliberty** [comment]
It looks like you are loading the nangate45 library not those from your new pdk.

**Reply by 706697916** [reply]
Well, I do not change the PDK name, but I changed the code details of PDK and all the relevant codes.

**Comment by maliberty** [comment]
The first warning is from yosys.  I know we've seen similar when you have un-quoted expression in the .lib.  If you don't care about those cells you could just ignore it.

For STA-123 can you give the full line it is complaining about?

For the last one - does your process have a layer called metal1?  I'm guessing you didn't update setRC.tcl

**Reply by 706697916** [reply]
The line 2601 2605 2609 reports STA-123.
You are right. I did not update the setRC.tcl in my platform.
![1683821141507](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/70256434/eee846cc-6b5c-4de6-87d9-daadf7691537)

**Comment by maliberty** [comment]
We don't support ecsm so resolving those warnings won't make much difference.  For now we only support NLDM and in the future will support CCS.

**Reply by maliberty** [reply]
In NLDM values is a complex attribute and here it is a simple attribute which is confusing the parser.  I think this is just a symptom of the lack of ecsm support.

**Reply by 706697916** [reply]
Thank you very much! 
For metal1, it is called M1 in my platform. It is not easy for me to change all metal1 to M1 in all OpenROAD-flow-scripts. I think I need to change M1 to metal1 in all my PDK files.

**Reply by rovinski** [reply]
ORFS does not have any layer names hard-coded anywhere. It is only in the platform files. Those files should be limited to: setRC.tcl (parasitics estimation), PDN config, and the KLayout technology file.

**Reply by 706697916** [reply]
That sounds great! I will have a try.

**Reply by 706697916** [reply]
I met another problem that I cannot handle. How to generate a lyt file?
As your tutorial mentioned, I need to generate a lyp file first. But I cannot get it by following the tutorial.
![1684119378453](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/70256434/8ea0eb01-23ac-4319-9333-be75c5d0b944)

**Reply by 706697916** [reply]
Besides, the route iteration takes very long time. How to reduce the numbers of iteration or release convergence?
<img width="980" alt="1684125217020" src="https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/70256434/0b313a87-672f-48b6-86cf-b6fba9552715">

**Reply by 706697916** [reply]
I cancelled the use of lyt file. Then the error occured.
<img width="1033" alt="1684131536250" src="https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/70256434/9c1ff377-f5c2-4a78-90fa-fa876bc3d70e">

**Reply by openroadie** [reply]
For routing taking too long... have you considered that the chip may be too congested and starting with a larger floor-plan ?

**Reply by 706697916** [reply]
How to change to a larger floor-plan?

**Comment by maliberty** [comment]
Have you looked at what is in the lyp at the line given?  There is no way to guess from a screen shot.

**Reply by 706697916** [reply]
Thank you, I have solved it.

**Comment by maliberty** [comment]
For routing you should load the drc errors and take a look at where they are occurring.

**Reply by 706697916** [reply]
There are so many optimization iterations, and it takes so much time. I wonder how to reduce the iterations.
<img width="970" alt="1689945882153" src="https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/70256434/8fa7f2ca-9522-4704-8534-0961cbbf3d06">

**Reply by maliberty** [reply, marked as answer]
Try setting the envar DETAILED_ROUTE_ARGS to "-droute_end_iter 5" to get it to stop earlier.

**Reply by 706697916** [reply]
Thank you, it works! 
Setting the envar DETAILED_ROUTE_ARGS to "-droute_end_iter 5" will get it to stop after 5th optimization iteration.
